### PR TITLE
Fix AppDomain.SetPrincipalPolicy bug for new threads

### LINKED ISSUE #31717: AppDomain.SetPrincipalPolicy(PrincipalPolicy.WindowsPrincipal) works only once
Setting the PrincipalPolicy on the current AppDomain to WindowsPrincipal works only for the first thread being started.  Any subsequent thread has Thread.CurrentPrincipal evaluated to NULL.

See the following example:

```
    internal static class Program 
    {
        private static void Main()
        {
            // Set the principal policy to WindowsPrincipal.
            AppDomain.CurrentDomain.SetPrincipalPolicy(PrincipalPolicy.WindowsPrincipal);
            
            // All threads should now have a Windows principal
            // according to the documentation on https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.setprincipalpolicy?view=netcore-3.1
            // Works here.
            var t = new Thread(PrintPrincipalInformation);
            t.Start();
            t.Join();

            //But doesn't work here anymore
            t = new Thread(PrintPrincipalInformation);
            t.Start();
            t.Join();

            // Create a principal to use for new threads.
            IIdentity identity = new GenericIdentity("NewUser");
            IPrincipal principal = new GenericPrincipal(identity, null);
            AppDomain.CurrentDomain.SetThreadPrincipal(principal);
            
            // Should create a new thread with the principal created above, 
            // according to the documentation here: https://docs.microsoft.com/en-us/dotnet/api/system.appdomain.setthreadprincipal?view=netcore-3.1
            // Also doesn't work.
            t = new Thread(PrintPrincipalInformation);
            t.Start();
            t.Join();
        
            // Wait for user input before terminating.
            Console.ReadLine();
        }

        static void PrintPrincipalInformation()
        {
            var curPrincipal = Thread.CurrentPrincipal;
            if(curPrincipal != null)
            {
                Console.WriteLine("Type: " + curPrincipal.GetType().Name);
                Console.WriteLine("Name: " + curPrincipal.Identity.Name);
                Console.WriteLine("Authenticated: " +
                                  curPrincipal.Identity.IsAuthenticated);
                Console.WriteLine();
                return;
            }
            Console.WriteLine("Thread.CurrentPrincipal is NULL");
        }
    }
```

.NET Core Runtime 3.1.1
.NET Core SDK 3.1.101

### PR DESCRIPTION
**Pull request by MarcoRossignoli**
Fixes https://github.com/dotnet/runtime/issues/31717

Old code did 
```cs
return CurrentThread._principal ?? (CurrentThread._principal = AppDomain.CurrentDomain.GetThreadPrincipal());
```
I missed to bring principal null check 😡 

cc: @jkotas @stephentoub 

PS. Debug cross repo is great(lib->coreclr), but it's a bit slow due to debug build of EE, is there a way to build EE in release and only managed part(`System.Private.CoreLib`) in debug?

### CODE REVIEW

**Comment by MarcoRossignoli** [comment]
Got this error from CI https://helix.dot.net/api/2019-06-17/jobs/c7ca5c04-d00f-417f-9b34-3e48d889129d/workitems/System.Threading.Tasks.Tests/console but seem unrelated to my update.

File: src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs [hunk 1]
```diff
@@ -133,7 +133,7 @@ public CultureInfo CurrentUICulture
         {
             get
             {
-                if (s_asyncLocalPrincipal is null)
+                if (s_asyncLocalPrincipal is null || s_asyncLocalPrincipal.Value is null)
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs]: This is going to call `AsyncLocal.Value` twice in the common case - once here, and second time below. Would it be better to call it just once? Something like:

```
IPrincipal? principal = s_asyncLocalPrincipal?.Value;
if (principal == null)
{
    CurrentPrincipal = (principal = AppDomain.CurrentDomain.GetThreadPrincipal());
}
return principal;
```

**Comment by jkotas** [comment]
The CI error is known problem. All PRs are hitting it.

File: src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs (hunk 1 above)
Comment (MarcoRossignoli) [reply src/libraries/System.Private.CoreLib/src/System/Threading/Thread.cs]: Yep will do

**Comment by jkotas** [comment]
> is there a way to build EE in release and only managed part

https://github.com/dotnet/runtime/blob/master/docs/workflow/building/libraries/README.md

**Comment by MarcoRossignoli** [comment]
> https://github.com/dotnet/runtime/blob/master/docs/workflow/building/libraries/README.md

Not sure there is what I would...for clr cannot build EE in release and c# in debug I can only specify one -c.

**Review by jkotas — APPROVED** [review]
Thank you!

**Comment by jkotas** [comment]
The test failures are known issue #32126

**Pull request merged**